consensus/core/src/authority_node.rs

```diff
@@ -381,14 +383,35 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
         // For now ask dag state directly
         let blocks = self.dag_state.read().get_blocks(&block_refs);
 
-        // Return the serialised blocks
+        // Now check if an ancestor's round is higher than the one that the peer has. If yes, then serve
+        // that ancestor blocks up to 10
+        let all_ancestors = blocks
+            .iter()
+            .flatten()
+            .flat_map(|block| block.ancestors().to_vec())
+            .filter(|block_ref| highest_accepted_rounds[block_ref.author] < block_ref.round)
+            .take(MAX_ADDITIONAL_BLOCKS)
+            .collect::<Vec<_>>();
+
+        let mut ancestor_blocks = vec![];
+        if !all_ancestors.is_empty() {
+            ancestor_blocks = self.dag_state.read().get_blocks(&all_ancestors);
+        }
+
+        // Return the serialised blocks & the ancestor blocks
         let result = blocks
             .into_iter()
             .flatten()
             .map(|block| block.serialized().clone())
             .collect::<Vec<_>>();
 
-        Ok(result)
+        let result_ancestors = ancestor_blocks
+            .into_iter()
+            .flatten()
+            .map(|block| block.serialized().clone())
+            .collect::<Vec<_>>();
+
+        Ok((result, result_ancestors))
```

Comment:
I'm have a pending change to chunk the response. Or we can add logic to limit the total response size. They will be easier if all blocks are concatenated together. Is that preferable than returning two separate `Vec<Bytes>`?

Reply:
@mwtian returning two different blocks makes things easier to validate & handle the response. So I would prefer returning two separate vectors. If it is a blocker/issue I can refactor the current struct would be my preference.

Reply:
I think by just returning one `Vec`, it should still straightforward to find the split point between requested blocks and missing ancestors? The main issue is if we need to implement response size limiting or chunking, will it become harder with two separate Vec of blocks?